### ISSUE TITLE
LSF executor plugin

### ISSUE THREAD

**Issue by BEFH** [snakemake/snakemake#2515]
**Is your feature request related to a problem? Please describe.**
It would be nice to have an LSF executor plugin like the `--slurm` executor. 

**Describe the solution you'd like**
I have forked the LSF plugin and done most of the work for the new plugin, but it needs some testing and needs a jobstep implemented. The repo is [here.](https://github.com/BEFH/snakemake-executor-plugin-lsf) 

**Describe alternatives you've considered**
There's also a generic executor, but that is sub-optimal.

**Comment by BEFH** [comment]
@johanneskoester @fgvieira, is there any way you could take a look at this?

**Comment by johanneskoester** [comment]
Merry Christmas @BEFH! Thanks a lot for your work on this! I won't have the time to take a look until mid of January (vacation now, sorry). However, the beaty of the plugin approach is that you can just release this on your own. I would even suggest to do that as early as possible, in order to get feedback by users. It will automatically appear in the snakemake plugin catalog once it is released to pypi.

**Comment by tgbrooks** [comment]
I would appreciate this and can test it for you. I'm interested in a replacement for the LSF profile that no longer works on snakemake version 8+. I tried installing this via `pip install snakemake==8.1.0 git+https://github.com/BEFH/snakemake-executor-plugin-lsf.git` on a fresh python 3.11.2 environment.

I get the following error from pip:

```
ERROR: Cannot install snakemake-executor-plugin-lsf==0.1.1 and snakemake==8.1.0 because these package versions have conflicting dependencies.

The conflict is caused by:
    snakemake 8.1.0 depends on snakemake-interface-executor-plugins<9.0 and >=8.1.3
    snakemake-executor-plugin-lsf 0.1.1 depends on snakemake-interface-executor-plugins<8.0.0 and >=7.0.0

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
```

Similarly for `snakemake==8.0.0`. I guess that makes sense since version 8 came out after this PR was made.

**Comment by BEFH** [comment]
Thanks! The plugin still needs a little work to be functional, even beyond
this. I'm going to be working on this a bit in the next few weeks, but
would also appreciate any help I can get.

**Comment by BEFH** [comment]
@johanneskoester hope things are going well. I have updated the executor for the most part, but I am still confused about the jobstep executor. Like Slurm, we have to submit LSF jobs one job at a time, preferably with a short pause inbetween submissions, but I can't figure out what the Slurm jobstep executor is doing. Could you please provide some advice before I publish the package on pip and conda?

**Comment by cmeesters** [comment]
Ah, sorry, just screening older issues

> but I can't figure out what the Slurm jobstep executor is doing.

My LSF experience is a bit rusty, but I don't think you need a sophisticated jobstep executor for LSF at all. Unter SLURM we trigger tasks with `srun`. There is no such equivalent under LSF (as far as I recall). SLURM users may use `srun` as the MPI-starter, too. Under LSF users would need to supply `mpiexec` (or similar) as the `mpi` resource in a rule. 

So, basically, under LSF you can just run your code and only need to catch the MPI case as we do in the slurm-executor.  

As for printing debug messages, does `self.logger.debug(<message>)` do the trick within the executors? It should work. The caveat for jobs, is, of course, that you need to wait until the batch system starts writing its buffer to a job log.

**Comment by BEFH** [comment]
I'll see about removing jobstep, thanks.

As for triggering debug messages, you can put in logger.debug() functions
all you want, but they won't print unless the logger level is set to debug.
My question is how to set the logger level to debug in order to see those
messages.

Do I need to do that in the executor code or is there a universal method in
Snakemake?

**Comment by cmeesters** [comment]
Right, you need to trigger with `--verbose`.

**Comment by BEFH** [comment]
Perfect thanks

**Comment by cmeesters** [comment]
BTW just to avoid misunderstandings

> I'll see about removing jobstep, thanks.

I do not know whether this is meant to drop the idea of a jobstep executor entirely. No, you still need it for the MPI case to be compatible with the resource specifications, if a rule specifies

```YAML
set-resources:
   rule:
       mpi: "mpirun -np $LSB_DJOB_NUMPROC"
```

If you like to, I can walk you through the SLURM code and its motivations (e.g. video call). But I do not know how to create an LSF cluster for the CI.

**Comment by BEFH** [comment]
I finally have some time if you're available to meet. How can we set up a meeting?

**Comment by cmeesters** [comment]
Sure thing! It's just that the last few weeks have been super stressful, and I am taking a week off. Thereafter, I am going to conduct an RDM course. So, the soonest would be Thursday, 30. May or Friday, 31. May. Where can I find your email?

**Comment by BEFH** [comment]
@cmeesters, ich habe sie nachgesucht und eine Ihnen eine E-mail geschickt

**Comment by BEFH** [comment]
I've made some updates in my executor based on our conversation

**Comment by benjeffery** [comment]
Thanks for your work on this plugin - I'm using it successfully and it is a great speed improvement over the old `--cluster` way.  I had an issue to raise (more than happy to PR a fix but wanted to discuss first) but I note that issues are disabled on the plugin repo. Was this deliberate?

**Comment by BEFH** [comment]
@benjeffery I just enabled issues, so feel free to open one. As for the pull request, I have an open pull request of my own that has a fair bit of refactoring, so if you open a pull request of your own, it may make sense to start with that branch.

**Comment by Wytamma** [comment]
Closing as [snakemake-executor-plugin-lsf](https://github.com/BEFH/snakemake-executor-plugin-lsf) is alive and well
